kpt-backstage-plugins/plugins/cad/src/components/ResourceEditorDialog/components/FirstClassEditors/ApplyReplacementsEditor/ApplyReplacementsEditor.tsx

```diff
@@ -0,0 +1,471 @@
+      );
+    }
+
+    return undefined;
+  };
+
+  const createResourceState = (): State => ({
+    name: resourceYaml.metadata.name,
+    annotations: resourceYaml.metadata.annotations ?? {},
+    labels: resourceYaml.metadata.labels ?? {},
+  });
+
+  const createReplacementState = (
+    resourceSelector?: ReplacementResourceSelector,
+    fieldPath?: string,
+    replacementOptions?: ReplacementOptions,
+  ): ReplacementState => ({
+    resourceId: findResource(resourceSelector)?.id ?? '',
+    resourceKind: resourceSelector?.kind ?? '',
+    resourceName: resourceSelector?.name ?? '',
+    fieldPath: fieldPath ?? '',
+    replaceValueOption: replacementOptions?.delimiter
+      ? ReplaceFieldValueOption.PARTIAL
+      : ReplaceFieldValueOption.FULL,
+    delimiter: replacementOptions?.delimiter ?? '',
+    index: (replacementOptions?.index ?? '').toString(),
+  });
+
+  const firstReplacement = resourceYaml?.replacements?.[0];
+  const firstReplacementTarget = resourceYaml?.replacements?.[0].targets?.[0];
+
+  const [state, setState] = useState<State>(createResourceState());
+  const [sourceState, setSourceState] = useState<ReplacementState>(
+    createReplacementState(
+      firstReplacement?.source,
+      firstReplacement?.source?.fieldPath,
+      undefined,
+    ),
+  );
+  const [targetState, setTargetState] = useState<ReplacementState>(
+    createReplacementState(
+      firstReplacementTarget?.select,
+      firstReplacementTarget?.fieldPaths?.[0],
+      firstReplacementTarget?.options,
+    ),
+  );
+
+  const [expanded, setExpanded] = useState<string>();
+
+  const classes = useEditorStyles();
+
+  const sourcePathSelectItems: SelectItem[] = getPathSelectItems(
+    packageResources,
+    sourceState.resourceId,
+  );
+  const targetPathSelectItems: SelectItem[] = getPathSelectItems(
+    packageResources,
+    targetState.resourceId,
+  );
+  const targetIndexSelectItems: SelectItem[] = getIndexSelectItems(
+    packageResources,
+    targetState,
+  );
+
+  useEffect(() => {
+    const replacementSource: ReplacementSource = {
+      kind: sourceState.resourceKind,
+      name: sourceState.resourceName,
+      fieldPath: sourceState.fieldPath,
+    };
+
+    const replacementTarget: ReplacementTarget = {
+      select: {
+        kind: targetState.resourceKind,
+        name: targetState.resourceName,
+      },
+      fieldPaths: [targetState.fieldPath],
+    };
+
+    if (
+      targetState.replaceValueOption === ReplaceFieldValueOption.PARTIAL &&
+      targetState.delimiter
+    ) {
+      replacementTarget.options = {
+        delimiter: targetState.delimiter,
+        index: parseInt(targetState.index, 10),
+      };
+    }
+
+    resourceYaml.metadata.name = state.name;
+    resourceYaml.metadata.labels = state.labels;
+    resourceYaml.metadata.annotations = state.annotations;
+    resourceYaml.replacements = [
+      {
+        source: replacementSource,
+        targets: [replacementTarget],
```

Comment:
Are we supporting just one target in the UI? The KRM function takes an array of targets to apply to. Same question about target fieldpaths.

Reply:
Confirming the UI only supports a single target and field path today. This can be extended in the future.